### ISSUE TITLE
WrapHandlersInATransactionScope + Outbox

### ISSUE THREAD

**Issue by Tsjunne** [Particular/NServiceBus.NHibernate#269]
I have a setup using RabbitMQ + NHibernate persistence on Oracle.
I have a service that needs a distributed transaction to integrate with another legacy database.

I tried setting the transport transaction to TransactionScope, but this is not allowed by the RabbitMQ transport.
I tried to enable the 'WrapHandlersInATransactionScope' feature. I see the SQL statements to insert the business data in the logs together with the outbox SQL, but it looks like data saved to the NH session is not persisted when both the outbox and WrapHandlersInATransactionScope are enabled. The outboxrecord is created tho. Is this to be expected? 
Setting up a new transactionscope in a ITransportReceiveContext seems to lock the DB in some way, resulting in a timeout.

Is there a way to make this setup work?

**Comment by MarcinHoppe** [comment]
@Tsjunne Can you provide you endpoint configuration and your handler code storing data via the NH session that reproduce the problem?

**Comment by Tsjunne** [comment]
[NSBRepo.zip](https://github.com/Particular/NServiceBus.NHibernate/files/850194/NSBRepo.zip)

Some comments on the repro:
This is the actual scenario i'm working in, so it's inside a saga timeout. It is basicly the scheduler example that used to be on the docs.
When starting the repro on an empty database, the saga won't get persisted either, so you might want to disable outbox and the tx wrapping to get the saga in.

One more thing that struck me: It looks like the timeoutmanager starts to back off to 1 minute polls when it determines it has nothing to do. You can simulate that by disabling outbox, tx wrapper and commenting the save of the business data. The timeout in the saga is set to trigger every 5 seconds. It will do that for a while and then not trigger sooner than every minute.

**Comment by SzymonPobiega** [comment]
@Tsjunne thank you for the repro. I managed to pinpoint the problem(s). When using `WrapHandlersInATransactionScope` the sequence of events is as follows:
 * Outbox opens a new connection and a native ADO transaction
 * Wrapper creates a `TransactionScope`
 * Saga persister stores a saga. This is when `AdoNetWithDistributedTransactionFactory` is pinged and it detects the presence of transaction scope and attaches its volatile enlistment
 * Wrapper complete the `TransactionScope`. Enlistment notification is invoked while it should not because the connection and transaction are not enlisted. It causes the native ADO transaction to rollback
 * Outbox persister stores the outbox record. The native transaction attached to the session is rolled back but for some reason it does not prevent NHibernate from using it to store the record. The record is stored in an implicit transaction

Forcing NHibernate to use the `AdoNetTransactionFactory`

```
configurationProperties[Environment.TransactionStrategy] = typeof(AdoNetTransactionFactory).FullName;
```

solves the problem but now the `TransactionScope` is effectively ignored by NServiceBus so it won't allow you to use DTC to coordinate storing data between the two databases. 

Would it be an option to separate the legacy code to a separate handler and endpoint and add the outbox table to the legacy database? This way the DTC would not be necessary. Currently there is no way to use `TransctionScope` in the outbox mode.

**Comment by Tsjunne** [comment]
Thank for looking at the repro @SzymonPobiega 
Our concrete setup is like this:
* we are building a new system to gradually replace the legacy.
* The legacy has a series of identical DB's where serveral tenants are spread among different DB's
* The service doing the integration needs both the outbox and a connection to a DB based on a tenant ID

I managed to redesign our setup with some minor changes that is based on the multitenant sample from the docs. We had to move around some responsabilities to ensure consistency, but we managed. I think it basically ends up like the situation you suggest.

Is there any plan to support outbox+DTC in this context? Not that i'm inclined to depend on DTC, but in a legacy migration scenario, any non-obtrusive migration path is welcome :)

Looking at the codebase I could imagine some scenario's to plug this into NSB, but not without lots of custom hooks in both NSB and NHibernate (The infamous 'already part of a transaction' issue when combining TransactionScope and session.BeginTransaction() with NH+Oracle comes to mind)
